node/node.go

```diff
@@ -155,21 +155,36 @@ func NewNode(ctx context.Context,
 		DiscardABCIResponses: config.Storage.DiscardABCIResponses,
 	})
 
-	genDoc, err := genesisDocProvider()
+	// Get the genesisDoc hash from db if exists
+	genDocHash, err := stateDB.Get(genesisDocHashKey)
 	if err != nil {
 		return nil, err
 	}
 
-	err = genDoc.ValidateAndComplete()
+	csGenDoc, err := genesisDocProvider(genDocHash)
+	if err != nil {
+		return nil, err
+	}
+
+	err = csGenDoc.GenesisDoc.ValidateAndComplete()
 	if err != nil {
 		return nil, fmt.Errorf("error in genesis doc: %w", err)
 	}
 
-	state, err := loadStateFromDBOrGenesisDoc(stateStore, stateDB, genDoc)
+	state, err := loadStateFromDBOrGenesisDoc(stateStore, stateDB, csGenDoc)
 	if err != nil {
 		return nil, err
 	}
 
+	// Save the genDoc hash in the store if it doesn't already exist for future verification
+	if len(genDocHash) == 0 {
```

Comment:
if it is not 0 should we not compare it against `cs.Sha256Checksum` to make sure they are the same?

Reply:
Actually I think this is where Sergio suggested we always save it? If so then we should first check whether the existing one matches the new one we generate, and then always save it.

Reply:
> if it is not 0 should we not compare it against `cs.Sha256Checksum` to make sure they are the same?

We do, as per the new contract on `GenesisDocProvider`.

Reply:
> Actually I think this is where Sergio suggested we always save it?

No need to overwrite it if it's already there and matches the file, I think.